{"nbformat":4,"nbformat_minor":0,"metadata":{"colab":{"name":"1.Make.Gene.Database.ipynb","provenance":[]},"kernelspec":{"name":"python3","display_name":"Python 3"}},"cells":[{"cell_type":"code","metadata":{"id":"tB6JV0HrFsod","colab_type":"code","outputId":"c35a0856-278d-4939-e8b1-f5c6bc1386e6","executionInfo":{"status":"ok","timestamp":1583348652176,"user_tz":300,"elapsed":4846,"user":{"displayName":"Jim Curran","photoUrl":"https://lh3.googleusercontent.com/a-/AOh14GhTZDYoWDjVYt4cZVbh-jZ_FhMZOizXdfytTH8J=s64","userId":"00865777746874349714"}},"colab":{"base_uri":"https://localhost:8080/","height":211}},"source":["#for parsing Genbank files.  Does these things:\n","#finds and reads a gb file stored on local drive\n","#makes a list of gene orientations, names, start, end, size and a dummy variable\n","#makes a codon usage table, but it ignores start and stop codons\n","#plots codon numbers vs. codon frequencies -- should give a straight line\n","#allows searches for genes\n","\n","genomeFileName = '/content/gdrive/Shared drives/2.Python/Genomics/Data Files/P.fluorescens.sequence.gb'\n","genesname =      '/content/gdrive/Shared drives/2.Python/Genomics/Data Files/P.fluorescens.sequence.Gene.Database.txt'\n","\n","import numpy as np\n","from matplotlib import pyplot as plt\n","from google.colab import drive\n","\n","drive.mount('/content/gdrive')\n","with open(genomeFileName, 'r') as gbfile:\n","  data = gbfile.read()\n","gbfile.close()\n","txtfile = str(data)\n","start = 3\n","stop = 3\n","frame = 0\n","\n","# COLLECTS ALL THE GENES IN A LIST CALLED \"features\" AND THE SEQUENCE WILL BE IN split2[1]\n","\n","split1 = txtfile.split('FEATURES') # split1[0] is annotations, split1[1] is the good stuff\n","split2 = split1[1].split('ORIGIN')  # gene locations are in split2[0] and genome sequence is in split2[1]\n","features = split2[0].split('  gene  ')  # separates each gene location and puts them in a list\n","\n","seq = ''\n","for char in split2[1]:                  #puts sequence in a single, long string\n","    if char in ('t', 'g', 'c', 'a'):\n","        seq += char\n","        \n","print('The number of genes in \"Features\" =', len(features) -1)\n","print('The \"sub-1\" Features line starts with:', features[1][:100])\n","print('The first 100 nucleotides =', seq[:100])        \n","\n","#COLLECT ALL THE FORWARD AND REVERSE GENES and labels them as 'FOR' or 'REV'\n","names =[]\n","forward =0\n","reverse =0\n","\n","for x in range(1, len(features)):\n","  if 'join' not in features[x]:\n","    if 'complement' not in features[x]: genedirection = 'FOR'\n","    elif 'complement' in features[x]:   genedirection = 'REV'\n","    cleanfeatures = features[x].replace('complement(','').replace(')', '')\\\n","    .replace('/', '').replace(' ', '').replace('gene=\"', '').replace('\"', '')\n","    workgene = cleanfeatures.split('\\n')\n","    genename = workgene[1]\n","    genelocation = workgene[0].split('.')\n","    genestart = genelocation[0] \n","    geneend = genelocation[2] \n","    genedata = [genedirection, genename, genestart.replace('>', '').replace('<', ''), geneend.replace('<', '').replace('>',''), 0, 0.0]\n","    names.append(genedata)\n","              \n","#converts gene starts and ends to integers.  Calulates gene length\n","print('The number of annotated genes =', len(names))\n","for y in range(len(names)):\n","  try:\n","    names[y][2] = int(names[y][2])\n","    names[y][3] = int(names[y][3])\n","    names[y][4] = int(names[y][3]) - int(names[y][2])\n","  except: print('Bad gene:', names[y])\n","                  \n","genes =[]\n","for x in range(len(names)):\n","  if names[x][0] == 'FOR':\n","    cds = seq[names[x][2] -1 : names[x][3]]\n","  elif names[x][0] == 'REV':\n","    transposed = ''\n","    flipped = seq[names[x][2] -1 : names[x][3]]\n","    for char in flipped:\n","      if char == 'a': newchar = 't'\n","      elif char == 't': newchar = 'a'\n","      elif char == 'c': newchar = 'g'\n","      elif char == 'g': newchar = 'c'\n","      transposed += newchar\n","    cds = transposed[::-1]\n","  genes.append(cds)\n","\n","#collects only those sequences that have proper start codons\n","startgenes =[]\n","startnames =[]\n","for x in range(len(genes)):\n","  firstcodon = genes[x][0:3]\n","  if firstcodon in ('atg', 'gtg', 'ctg', 'ttg', 'att', 'ata', 'aaa'):\n","    startgenes.append(genes[x])\n","    startnames.append(names[x])\n","\n","#Collects only those sequences that do not contain internal stop codons\n","goodgenes =[]\n","goodnames =[]\n","for x in range(len(startgenes)):\n","  delete = False\n","  for a in range(start, len(startgenes[x]) - stop):\n","    if a % 3 == frame and delete == False:\n","      codon = startgenes[x][a:a+2] \n","      if codon in ('tag', 'taa', 'tga'):\n","        delete = True\n","  if delete == False : \n","    goodgenes.append(startgenes[x])\n","    goodnames.append(startnames[x])    \n","\n","print(goodnames[0])\n","print(goodgenes[0])\n","\n","outgenes =[]\n","for x in range(len(goodgenes)):\n","  gene = goodnames[x][1] +','+ str(goodnames[x][2]) +','+ str(goodnames[x][3]) +','+  goodgenes[x] + '\\n'\n","  outgenes.append(gene)\n","\n","with open(genesname, 'w') as g:\n","  for x in range(len(outgenes)):\n","    g.write(outgenes[x])\n","g.close()\n","\n","print('Done.')\n"," "],"execution_count":2,"outputs":[{"output_type":"stream","text":["Drive already mounted at /content/gdrive; to attempt to forcibly remount, call drive.mount(\"/content/gdrive\", force_remount=True).\n","The number of genes in \"Features\" = 6085\n","The \"sub-1\" Features line starts with:           563..2089\n","                     /gene=\"dnaA\"\n","                     /locus_tag=\"PSF113_RS3023\n","The first 100 nucleotides = tgttacctggttcgtccacaacgggccggaatggcccccgttttaagagaccggggattctagagaaagcaagcctctaggtcaatttccaaccagcgtt\n","The number of annotated genes = 6084\n","['FOR', 'dnaA', 563, 2089, 1526, 0.0]\n","gtgtcagtggaactttggcagcagtgcgtagagcttttgcgcgatgagctgcctgcccaacaattcaacacttggatccgcccactacaggtcgaagccgaaggcgacgagttgcgtgtctatgcaccgaaccgtttcgttctcgactgggtcaacgaaaagtacctgggccgggtcatggaactgctcgacgagcatggcaatggcatggcgcctgcgctgtccttattaataggcagcaagcgcagctcggcaccgcgcgccgcacccaacgcgccgctggccgccgcagcgtcacaggcccagcagactgcaacggcaccggttaacaaccatgccgccccggcacccgcgcctgccccgagcaaacgcaacgcacaaaaggtcgccgaggtcagcgaagaaccctcccgggacagcttcgacccgatggccggtgccagctcccagcaggcgccggtgcgtgccgaacagcgcaccgtccaggtcgaaggtgcgctcaagcacaccagctacctgaaccggacctttaccttcgagaatttcgtcgaaggtaaatccaaccaactggctcgcgctgcggcctggcaggtggcggacaaccccaagcatggctacaacccactcttcctttatggcggcgtaggcttgggtaagacccacttgatgcatgcggtgggcaaccatctattaaagaagaacccgaatgccaaggtcgtgtacctgcattccgagcgttttgtggccgacatggtcaaggcgctgcaactgaacgcgatcaacgagttcaagcgtttctaccgttcggtggacgccttgctgatcgacgacattcagttcttcgcccgcaaggaacgttcccaggaagagtttttccacaccttcaacgcattgcttgaaggcgggcagcaggtcattctcaccagtgaccgctacccgaaagaaatcgaaggccttgaagaacgcctcaagtcccgtttcggctggggtctgacggtggccgtcgagccgccggagctggaaacccgcgtcgcgatcctgatgaagaaagccgaccaggccaaggtcgagctgcctcacgacgcggcgttcttcatcgcccagcgtatccgctccaacgtgcgtgagctcgaaggtgccctcaagcgggtgatcgctcactcgcacttcatgggccgcgacatcaccatcgagttgattcgcgaatccttgaaggacctgctggcgctccaggacaaattggtttctgtggataacattcagcggacggttgcggaatactacaaaatcaagatttccgacctgctgtccaaacgccgttcgcgttcggtagcacgtccgcgtcaggtggcgatggccctgtccaaggaactgaccaaccacagcctgccggaaatcggcgatgtgtttggcggacgcgaccacaccacggtgttgcacgcctgccgcaagatcaacgaacttaaggaatccgacgcggacatccgcgaggactacaagaacctgctgcgtacactgaccacttga\n","Done.\n"],"name":"stdout"}]}]}