{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "name": "7.Map.Genes.to.Contig.ipynb",
      "provenance": []
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    }
  },
  "cells": [
    {
      "cell_type": "code",
      "metadata": {
        "id": "ZTkSMekMHvbb",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 364
        },
        "outputId": "639d7ece-3fff-4653-b761-41583091b99b"
      },
      "source": [
        "# for determining whether genes from a reference database occur in a contig\n",
        "\n",
        "refgenome = 'E.coli.K12'\n",
        "contig = 'Sample.Genome.Contig.'\n",
        "passnum = 1\n",
        "limit = 0.95          # the fraction of bases that must match\n",
        "#step = the number of genes in a step size between search. It is assigned just prior to the first search routine \n",
        "#numzoom = the number of genes in the drill-down.  It is defined just above the second search routine\n",
        "#import requests\n",
        "#!pip install fuzzywuzzy\n",
        "#from fuzzywuzzy import fuzz\n",
        "import math\n",
        "from google.colab import drive\n",
        "drive.mount('/content/gdrive')\n",
        "contigname = '/content/gdrive/Shared drives/2.Python/Genomics/Data Files/' + contig + str(passnum)+ '.txt'\n",
        "genesname =  '/content/gdrive/Shared drives/2.Python/Genomics/Data Files/' + refgenome + '.Gene.Database.txt'\n",
        "outfile =    '/content/gdrive/Shared drives/2.Python/Genomics/Data Files/' + contig + refgenome + '.MappedGenes.txt'\n",
        "\n",
        "with open(contigname, 'r') as f:\n",
        "  contigfile = str(f.read())\n",
        "  #print(repr(data)[:200])\n",
        "  #print(repr(data)[-200:])\n",
        "f.close()\n",
        "contig = contigfile.replace('\"', '').replace('b', '')\n",
        "print('Contig length =', len(contig))\n",
        "with open(genesname, 'r') as g:\n",
        "  genesfile = str(g.read())\n",
        "  #print(repr(data)[:200])\n",
        "  #print(repr(data)[-200:])\n",
        "g.close()\n",
        "genelist = genesfile.split('\\n')\n",
        "genelist.pop()\n",
        "print('The database has', len(genelist), 'genes.')\n",
        "\n",
        "genes = []\n",
        "for x in range(len(genelist)):\n",
        "  row = genelist[x].split(',')\n",
        "  genes.append(row)\n",
        "\n",
        "#print(genes[0])\n",
        "#print(contig[:100])\n",
        "\n",
        "def rev(string1):  \n",
        "  string2 =''\n",
        "  for x in string1:\n",
        "    if   x == 'c': string2 += 'g'\n",
        "    elif x == 'g': string2 += 'c'\n",
        "    elif x == 't': string2 += 'a'\n",
        "    elif x == 'a': string2 += 't'\n",
        "  string3 = string2[::-1]  \n",
        "  return(string3)\n",
        "\n",
        "def match(string1, string2):\n",
        "  sum =0\n",
        "  for x in range(len(string1)):\n",
        "    if string1[x] == string2[x]:\n",
        "      sum +=1\n",
        "  ratio = sum / len(string1)\n",
        "  return(ratio)\n",
        "\n",
        "hits =[]\n",
        "stop = False\n",
        "step = math.floor(len(contig) / 1500 )   # determines how many genes between the initial search\n",
        "print('Size of the step in the inital pass through the database =', step)\n",
        "numzoom = math.ceil(len(contig) / 600 )   # determines how many genes to drill down on\n",
        "print('Number of genes searched on either side of the initial hit =', numzoom)\n",
        "print()\n",
        "for x in range(3000, len(genes)):\n",
        "  if x % step == 0 and stop == False :\n",
        "    if len(genes[x][3]) in range (50, len(contig) -50):\n",
        "      #print(genes[x][0])\n",
        "      query = genes[x][3][9:-9]\n",
        "      revsq = rev(genes[x][3][9:-9])\n",
        "      sum =0\n",
        "      if x % 100 == 0: print('Number of genes searched thus far...', x)\n",
        "      for y in range(len(contig) - len(query)):\n",
        "        window = contig[y: y + len(query)]\n",
        "        if match(query, window) >= limit or match(revsq, window) >= limit:  # can use fuzz.ratio instead of match\n",
        "        #if fuzz.ratio(query, window) >= limit or fuzz.ratio(revsq, window) >= limit:  # can use fuzz.ratio instead of match\n",
        "          hits.append( ['Gene #', x, genes[x][0], genes[x][1] ]  )\n",
        "          print('Gene #', x, ',', genes[x][0], ', at position:', genes[x][1])\n",
        "          stop = True\n",
        "          new = x\n",
        "          break\n",
        "\n",
        "for z in range(new - numzoom, new + numzoom):\n",
        "  if z != new:\n",
        "    if len(genes[z][3]) in range (50, len(contig) -50):\n",
        "      query = genes[z][3][9:-9]\n",
        "      revsq = rev(genes[z][3][9:-9])\n",
        "      sum =0\n",
        "      for y in range(len(contig) - len(query)):\n",
        "        window = contig[y: y + len(query)]\n",
        "        if match(query, window) >= limit or match(revsq, window) >= limit:  # can use fuzz.ratio instead of match\n",
        "        #if fuzz.ratio(query, window) >= limit or fuzz.ratio(revsq, window) >= limit:  # can use fuzz.ratio instead of match\n",
        "#          hits.append( ['Gene number: ',z,  genes[z][:3] ] )\n",
        "          hits.append( ['Gene #', z, genes[z][0], genes[z][1] ]  )\n",
        "          print('Gene #', z, ',', genes[z][0], ', at position:', genes[z][1])\n",
        "          break\n",
        "\n",
        "print()\n",
        "hits.sort(key = lambda x: x[1]) \n",
        "outlist =[]\n",
        "for x in range(len(hits)):\n",
        "  outlist.append( [ hits[x][0] +','+ str(hits[x][1]) +','+ str(hits[x][2]) +','+ str(hits[x][3]) +'\\n' ] )\n",
        "  \n",
        "print()\n",
        "with open(outfile, 'w') as h:\n",
        "  for x in range(len(hits)):\n",
        "    outline = hits[x][0] +','+ str(hits[x][1]) +','+ str(hits[x][2]) +','+ str(hits[x][3])\n",
        "    h.write(outline + '\\n')\n",
        "    print(outline)\n",
        "h.close()\n",
        "print()\n",
        "print('Hits for reference genome,', refgenome, 'with a match score >=', limit, '. These gene hits are not saved to a file.')\n",
        "print('Done.')"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "Drive already mounted at /content/gdrive; to attempt to forcibly remount, call drive.mount(\"/content/gdrive\", force_remount=True).\n",
            "Contig length = 6780\n",
            "The database has 4175 genes.\n",
            "Size of the step in the inital pass through the database = 4\n",
            "Number of genes searched on either side of the initial hit = 12\n",
            "\n",
            "Number of genes searched thus far... 3000\n",
            "Gene # 3020 , rbfA , at position: 3312777\n",
            "Gene # 3021 , infB , at position: 3313342\n",
            "Gene # 3022 , nusA , at position: 3316039\n",
            "Gene # 3023 , rimP , at position: 3317554\n",
            "\n",
            "\n",
            "Gene #,3020,rbfA,3312777\n",
            "Gene #,3021,infB,3313342\n",
            "Gene #,3022,nusA,3316039\n",
            "Gene #,3023,rimP,3317554\n",
            "\n",
            "Hits for reference genome, E.coli.K12 with a match score >= 0.95 . These gene hits are not saved to a file.\n",
            "Done.\n"
          ],
          "name": "stdout"
        }
      ]
    }
  ]
}